### PR TITLE
[Profiler] add config option to remove 'Call stack' field from trace file

### PR DESCRIPTION
**Pull request by slgong-fb**
Summary: `Call stack` field increases trace file size exponentially for Python stack tracing (need to be deprecated carefully). Added a config option to avoid this increase.

Test Plan:
`experimental_config=_ExperimentalConfig(no_callstack_trace=True),` will remove the field.
+ CI tests

Differential Revision: D39489828

### CODE REVIEW

File: torch/_C/_profiler.pyi [hunk 1]
```diff
@@ -35,6 +35,7 @@ class _ExperimentalConfig:
         self,
         profiler_metrics: List[str] = ...,
         profiler_measure_per_kernel: bool = ...,
+        no_callstack_trace: bool = ...,
```
Comment (robieta) [inline torch/_C/_profiler.pyi:38]: I really like the idea of putting this flag in ExperimentalConfig. I think we should make it more generic. Maybe `new_trace_format`? (@aaronenyeshi, @chaekit feel free to bikeshed.)

File: torch/csrc/autograd/profiler_kineto.cpp [hunk 2]
```diff
@@ -85,8 +86,12 @@ struct MetadataBase {
     }
   }
 
-  void addMetadata(const std::string& key, const std::string& value) {
-    if (kineto_activity_ && !value.empty() && value != "\"\"") {
+  void addMetadata(
+      const std::string& key,
+      const std::string& value,
+      const bool no_callstack_trace = false) {
+    if (!no_callstack_trace && kineto_activity_ && !value.empty() &&
```
Comment (robieta) [inline torch/csrc/autograd/profiler_kineto.cpp:93]: `addMetadata` is a generic method. Config based omission should be handled by the caller.

File: torch/csrc/autograd/profiler_kineto.cpp [hunk 3]
```diff
@@ -102,12 +107,16 @@ struct MetadataBase {
...
       const std::shared_ptr<Result>& result,
-      KinetoEvent& kineto_event)
+      KinetoEvent& kineto_event,
+      const bool no_callstack_trace)
       : MetadataBase(result) {
     result->visit(*this);
     const auto module_hierarchy = kineto_event.moduleHierarchy();
     addMetadata("Module Hierarchy", stacksToStr(module_hierarchy.vec(), "."));
-    addMetadata("Call stack", stacksToStr(kineto_event.stack().vec(), ";"));
+    addMetadata(
```
Comment (robieta) [inline torch/csrc/autograd/profiler_kineto.cpp:116]: This would also drop call stack from JIT, which we want to preserve. (Since the end goal is to flip a switch.)

**Review by robieta — CHANGES_REQUESTED** [review]
High level thought: perhaps the simplest thing to do would be to change the signature of `KinetoEvent::stack()` to `KinetoEvent::stack(bool jit_only=false)`. Then you could just pass the config value to `stack` when generating metadata.

File: torch/csrc/autograd/profiler_kineto.cpp (hunk 2 above)
Comment (slgong-fb) [reply torch/csrc/autograd/profiler_kineto.cpp]: okay

File: torch/csrc/autograd/profiler_kineto.cpp (hunk 3 above)
Comment (slgong-fb) [reply torch/csrc/autograd/profiler_kineto.cpp]: I changed to drop call stack info is KinetoEvent constructor, where `python_stack_` is built, and changes are working fine in my local test.

File: torch/_C/_profiler.pyi (hunk 1 above)
Comment (slgong-fb) [reply torch/_C/_profiler.pyi:38]: will mark resolved unless we are trying `new_trace_format` in this PR.

File: torch/csrc/autograd/profiler_kineto.cpp [hunk 4]
```diff
@@ -618,16 +620,19 @@ std::unique_ptr<ProfilerResult> disableProfiler() {
...
     : result_{result} {
   TORCH_INTERNAL_ASSERT(result != nullptr);
 
-  // Populate Python stack
-  auto parent = result_->parent_.lock();
-  while (parent != nullptr) {
-    parent->visit_if_base<PyExtraFieldsBase>(
-        [&](const auto& i) { python_stack_.push_back(parent->name()); });
-    parent = parent->parent_.lock();
+  if (!no_callstack_trace) {
```
Comment (robieta) [inline torch/csrc/autograd/profiler_kineto.cpp:628]: not no callstack trace is kind of hard to parse.

File: torch/_C/_profiler.pyi (hunk 1 above)
Comment (robieta) [reply torch/_C/_profiler.pyi:38]: I was mostly thinking just because that way we wouldn't have to update the name every time we want to improve the trace format. The code itself doesn't have to change. (Maybe `updated_format`?)

**Review by robieta — APPROVED** [review]
I had a couple closing comments, but nothing major. So if you want to just submit and follow up later you're welcome to.

File: torch/_C/_profiler.pyi (hunk 1 above)
Comment (slgong-fb) [reply torch/_C/_profiler.pyi:38]: Good point. I chose `verbose` (more general) for this PR changes instead of `no_callstack_trace`, with an opposite default value and comment. I guess we would update trace format in the future but not only just reducing it but improve in a different way?

Comment (robieta) [reply torch/_C/_profiler.pyi]: Ohh, I like `verbose`.
